balloc/balloc.c

```diff
@@ -69,40 +70,152 @@ struct balloc_allocation_context {
 	struct m0_ext                  bac_final;/*< final results */
 };
 
-static inline int btree_lookup_sync(struct m0_be_btree  *tree,
-			       const struct m0_buf *key,
-			       struct m0_buf       *val)
+static int btree_lookup_callback(struct m0_btree_cb  *cb,
+				 struct m0_btree_rec *rec)
 {
-	return M0_BE_OP_SYNC_RET(op, m0_be_btree_lookup(tree, &op, key, val),
-	                         bo_u.u_btree.t_rc);
+	struct m0_btree_rec     *datum = cb->c_datum;
+
+	/** Copy both keys and values. Keys are copied to cover slant case. */
+	m0_bufvec_copy(&datum->r_key.k_data, &rec->r_key.k_data,
+		       m0_vec_count(&rec->r_key.k_data.ov_vec));
+	m0_bufvec_copy(&datum->r_val, &rec->r_val,
+		       m0_vec_count(&rec->r_val.ov_vec));
+	return 0;
 }
 
-static inline int btree_insert_sync(struct m0_be_btree  *tree,
-			       struct m0_be_tx     *tx,
-			       const struct m0_buf *key,
-			       const struct m0_buf *val)
+static inline int btree_lookup_sync(struct m0_btree     *tree,
+				    const struct m0_buf *key,
+				    struct m0_buf       *val,
+				    bool                 slant)
 {
-	return M0_BE_OP_SYNC_RET(op,
-				 m0_be_btree_insert(tree, tx, &op, key, val),
-	                         bo_u.u_btree.t_rc);
+	struct m0_btree_op   kv_op  = {};
+	void                *k_ptr  = key->b_addr;
+	void                *v_ptr  = val->b_addr;
+	m0_bcount_t          ksize  = key->b_nob;
+	m0_bcount_t          vsize  = val->b_nob;
+	uint64_t             flags  = slant ? BOF_SLANT : BOF_EQUAL;
+
+	struct m0_btree_rec  rec    = {
+			    .r_key.k_data = M0_BUFVEC_INIT_BUF(&k_ptr, &ksize),
+			    .r_val        = M0_BUFVEC_INIT_BUF(&v_ptr, &vsize),
+			};
+	struct m0_btree_cb   get_cb = {.c_act = btree_lookup_callback,
+				       .c_datum = &rec,
+				      };
+
+	return M0_BTREE_OP_SYNC_WITH_RC(&kv_op,
+					m0_btree_get(tree, &rec.r_key, &get_cb,
+						     flags, &kv_op));
 }
 
-static inline int btree_update_sync(struct m0_be_btree  *tree,
-			       struct m0_be_tx     *tx,
-			       const struct m0_buf *key,
-			       const struct m0_buf *val)
+static int btree_insert_callback(struct m0_btree_cb  *cb,
```

Comment:
Can lookup and insert callbacks be combined into one, as they seems to be having same code ?

Reply:
The source and destination of copy is different for lookup and insert so we keep them separate.